In [24]:
# Імпортуємо бібліотеки
import pandas as pd
import numpy as np

In [23]:
# Завантажуємо і об'єднуємо файли в один
files = [
    "tz_opendata_z01012022_po01012023.csv",
    "tz_opendata_z01012023_po01012024.csv",
    "tz_opendata_z01012024_po01012025.csv"
]

dfs = [pd.read_csv(f, sep=";", encoding="utf-8") for f in files]
df_all = pd.concat(dfs, ignore_index=True)

print(df_all.shape)

# Виводимо перших 5 рядків
df_all.head()

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_10428\2151789856.py:8: DtypeWarning: Columns (17,18) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs = [pd.read_csv(f, sep=";", encoding="utf-8") for f in files]
C:\Users\ADMIN\AppData\Local\Temp\ipykernel_10428\2151789856.py:8: DtypeWarning: Columns (17,18) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs = [pd.read_csv(f, sep=";", encoding="utf-8") for f in files]
C:\Users\ADMIN\AppData\Local\Temp\ipykernel_10428\2151789856.py:8: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs = [pd.read_csv(f, sep=";", encoding="utf-8") for f in files]


(6215184, 20)


PERSON  REG_ADDR_KOATUU  OPER_CODE  \
0      P     1.225284e+09        308   
1      P     3.220881e+09        254   
2      P     5.610700e+09        315   
3      P     8.036600e+09        315   
4      P     5.610100e+09        410   

                                           OPER_NAME       D_REG  DEP_CODE  \
0  ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ЗА ДОГ. КУПI...  21.12.2022     12537   
1                    НАЛЕЖНИЙ КОРИСТУВАЧ. РЕЄСТРАЦІЯ  07.09.2022     10000   
2  ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВ. ВЛАСН. ПО ДОГОВОРУ У...  09.12.2022     12286   
3  ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВ. ВЛАСН. ПО ДОГОВОРУ У...  02.07.2022     12294   
4          ПЕРЕРЕЄСТРАЦІЯ ПРИ ЗАМІНІ НОМЕРНОГО ЗНАКУ  13.09.2022     12337   

                    DEP       BRAND    MODEL                VIN  MAKE_YEAR  \
0              ТСЦ 1248         KIA     SOUL  KNAJT811AC7443261       2012   
1  OLD_ДДАІ МВС УКРАЇНИ      NISSAN  QASHQAI  SJNFBNJ10U2471634       2012   
2              ТСЦ 3247  VOLKSWAGEN     GOLF  WVWZZZ1KZCW268678       2012   
3              ТСЦ 8045     RENAULT  FLUENCE  VF1LZBB0647079727       2012   
4              ТСЦ 5641      NISSAN    TIIDA  3N1FCAC11UK561776       2012   

        COLOR      KIND       BODY    PURPOSE             FUEL  CAPACITY  \
0  КОРИЧНЕВИЙ  ЛЕГКОВИЙ  УНІВЕРСАЛ  ЗАГАЛЬНИЙ   БЕНЗИН АБО ГАЗ    1591.0   
1       БІЛИЙ  ЛЕГКОВИЙ    ХЕТЧБЕК  ЗАГАЛЬНИЙ   БЕНЗИН АБО ГАЗ    1997.0   
2       БІЛИЙ  ЛЕГКОВИЙ    ХЕТЧБЕК  ЗАГАЛЬНИЙ           БЕНЗИН    1390.0   
3       СІРИЙ  ЛЕГКОВИЙ      СЕДАН  ЗАГАЛЬНИЙ  ДИЗЕЛЬНЕ ПАЛИВО    1461.0   
4      ЧОРНИЙ  ЛЕГКОВИЙ    ХЕТЧБЕК  ЗАГАЛЬНИЙ           БЕНЗИН    1598.0   

  OWN_WEIGHT TOTAL_WEIGHT N_REG_NEW  
0     1217.0       1690.0  АЕ7265ХА  
1     1310.0       1960.0  АІ1954ЕМ  
2     1020.0       1850.0  АІ2805ОН  
3     1280.0       1680.0  КА1560ІЕ  
4     1268.0       1715.0  ВК2037ІВ

Повідомлення DtypeWarning. 
У різних файлах один і той самий стовпець має різні типи (наприклад, у одному число, у другому текст). У стовпцях 17 і 18 в різних частинах файлу pandas зустрів різні типи (наприклад, числа та текст).
У такому випадку pandas читає все як object (рядки). Тому далі виводимо усі типи даних і приводимо до правильного формату.

In [10]:
# Перевіряємо типи даних у злитому даатсеті
df_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6215184 entries, 0 to 6215183
Data columns (total 20 columns):
 #   Column           Dtype  
---  ------           -----  
 0   PERSON           object 
 1   REG_ADDR_KOATUU  float64
 2   OPER_CODE        int64  
 3   OPER_NAME        object 
 4   D_REG            object 
 5   DEP_CODE         int64  
 6   DEP              object 
 7   BRAND            object 
 8   MODEL            object 
 9   VIN              object 
 10  MAKE_YEAR        int64  
 11  COLOR            object 
 12  KIND             object 
 13  BODY             object 
 14  PURPOSE          object 
 15  FUEL             object 
 16  CAPACITY         float64
 17  OWN_WEIGHT       object 
 18  TOTAL_WEIGHT     object 
 19  N_REG_NEW        object 
dtypes: float64(2), int64(3), object(15)
memory usage: 948.4+ MB


**Висовок по результатах аналізу типів даних:**

Аналізуючи типи даних бачимо, ряд проблем:
- OWN_WEIGHT і TOTAL_WEIGHT мають тип даних object, а мають мати тип даних числовий. 
- OPER_NAME, DEP, BRAND, MODEL, COLOR, KIND, BODY, PURPOSE, FUEL - краще перевести до category. Бо вони не будуть обчислюватись, а тільки групуватись/фільтруватись. Це зменить об'єм пам'яті. 
- D_REG має тип даних object, краще перевести до datetime. 
Тому далі переводиво ці дві колонки до числового типу.

In [18]:
# Перетворення OWN_WEIGHT і TOTAL_WEIGHT у числовий формат
df_all["OWN_WEIGHT"] = pd.to_numeric(df_all["OWN_WEIGHT"], errors="coerce")
df_all["TOTAL_WEIGHT"] = pd.to_numeric(df_all["TOTAL_WEIGHT"], errors="coerce")

In [16]:
# Переводимо категоріальні дані до category
cat_columns = [
    "OPER_NAME", "DEP", "BRAND", "MODEL", "COLOR",
    "KIND", "BODY", "PURPOSE", "FUEL"
]
for col in cat_columns:
    df_all[col] = df_all[col].astype("category")

In [20]:
# Переводимо D_REG в datetime, щоб Tableau одразу коректно його розпізнав.
df_all["D_REG"] = pd.to_datetime(df_all["D_REG"], errors="coerce")

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_10428\3340925996.py:2: UserWarning: Parsing dates in %d.%m.%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_all["D_REG"] = pd.to_datetime(df_all["D_REG"], errors="coerce")


In [25]:
# Вказуємо одразу формат дати
df_all["D_REG"] = pd.to_datetime(df_all["D_REG"], format="%d.%m.%Y", errors="coerce")

In [66]:
# Кількість некоректно розпізнаних дат (NaT)
invalid_dates = df_all["D_REG"].isna().sum()
print(f"Кількість некоректних дат: {invalid_dates}")

# Переглянути приклади некоректних рядків
print(df_all[df_all["D_REG"].isna()][["D_REG"]].head(10))

# Мінімальна та максимальна дата
print("Мінімальна дата:", df_all["D_REG"].min())
print("Максимальна дата:", df_all["D_REG"].max())

# Скільки різних років є в даних
print("Унікальні роки:", df_all["D_REG"].dt.year.unique())

# Розподіл по роках (для наглядності)
print(df_all["D_REG"].dt.year.value_counts().sort_index())


Кількість некоректних дат: 4469276
        D_REG
1745908   NaT
1745909   NaT
1745910   NaT
1745911   NaT
1745912   NaT
1745913   NaT
1745914   NaT
1745915   NaT
1745916   NaT
1745917   NaT
Мінімальна дата: 2022-01-01 00:00:00
Максимальна дата: 2022-12-31 00:00:00
Унікальні роки: [2022.   nan]
D_REG
2022.0    1745908
Name: count, dtype: int64


In [22]:
# Перевіряємо оновлені типи даних у злитому даатсеті
df_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6215184 entries, 0 to 6215183
Data columns (total 20 columns):
 #   Column           Dtype         
---  ------           -----         
 0   PERSON           object        
 1   REG_ADDR_KOATUU  float64       
 2   OPER_CODE        int64         
 3   OPER_NAME        category      
 4   D_REG            datetime64[ns]
 5   DEP_CODE         int64         
 6   DEP              category      
 7   BRAND            category      
 8   MODEL            category      
 9   VIN              object        
 10  MAKE_YEAR        int64         
 11  COLOR            category      
 12  KIND             category      
 13  BODY             category      
 14  PURPOSE          category      
 15  FUEL             category      
 16  CAPACITY         float64       
 17  OWN_WEIGHT       float64       
 18  TOTAL_WEIGHT     float64       
 19  N_REG_NEW        object        
dtypes: category(9), datetime64[ns](1), float64(4), int64(3), object(3)
m

In [29]:
# Додаємо технічну колонку рік набору даних. Це допоможе в Tableau будувати зріз по роках.

files = {
    "tz_opendata_z01012022_po01012023.csv": 2022,
    "tz_opendata_z01012023_po01012024.csv": 2023,
    "tz_opendata_z01012024_po01012025.csv": 2024
}

dfs = []

for f, year in files.items():
    df_temp = pd.read_csv(
        f,
        sep=";",
        encoding="utf-8",
        dtype={
            "OWN_WEIGHT": "string",
            "TOTAL_WEIGHT": "string"
        },
        low_memory=False
    )
    df_temp["YEAR"] = year
    dfs.append(df_temp)

df_all = pd.concat(dfs, ignore_index=True)

# тепер приводимо вагу до числового типу
df_all["OWN_WEIGHT"] = pd.to_numeric(df_all["OWN_WEIGHT"], errors="coerce")
df_all["TOTAL_WEIGHT"] = pd.to_numeric(df_all["TOTAL_WEIGHT"], errors="coerce")

print(df_all.shape)
print(df_all.dtypes)


(6215184, 21)
PERSON              object
REG_ADDR_KOATUU    float64
OPER_CODE            int64
OPER_NAME           object
D_REG               object
DEP_CODE             int64
DEP                 object
BRAND               object
MODEL               object
VIN                 object
MAKE_YEAR            int64
COLOR               object
KIND                object
BODY                object
PURPOSE             object
FUEL                object
CAPACITY           float64
OWN_WEIGHT           Int64
TOTAL_WEIGHT         Int64
N_REG_NEW           object
YEAR                 int64
dtype: object


In [36]:
# Створюємо колонку вік авто на момент року реєстрації
df_all['CAR_AGE'] = df_all['YEAR'] - df_all['MAKE_YEAR']

# Якщо є негативні значення або пропуски — замінюємо на NaN
df_all['CAR_AGE'] = df_all['CAR_AGE'].apply(lambda x: x if x >= 0 else pd.NA)

# Перевірка
df_all[['MAKE_YEAR', 'YEAR', 'CAR_AGE']].head()

MAKE_YEAR  YEAR  CAR_AGE
0       2012  2022       10
1       2012  2022       10
2       2012  2022       10
3       2012  2022       10
4       2012  2022       10

In [43]:
# Створюємо категорії віку авто за роками
bins = [-1, 1, 5, 10, 20, 100]  # верхня межа 100 років
labels = ['0-1', '2-5', '6-10', '11-20', '21+']

df_all['AGE_CATEGORY'] = pd.cut(df_all['CAR_AGE'], bins=bins, labels=labels)

# Перевірка
df_all[['MAKE_YEAR', 'YEAR', 'CAR_AGE', 'AGE_CATEGORY']].head(10)

MAKE_YEAR  YEAR  CAR_AGE AGE_CATEGORY
0       2012  2022       10         6-10
1       2012  2022       10         6-10
2       2012  2022       10         6-10
3       2012  2022       10         6-10
4       2012  2022       10         6-10
5       2006  2022       16        11-20
6       2012  2022       10         6-10
7       2012  2022       10         6-10
8       2011  2022       11        11-20
9       2012  2022       10         6-10

In [41]:
# Аналіз пропущених значень
missing_data = df_all.isnull().sum()                        # Кількість пропусків
missing_percent = (missing_data / len(df_all)) * 100       # Відсоток пропусків

# Створюємо зручну таблицю
missing_summary = pd.DataFrame({
    'Пропущені': missing_data,
    '%': missing_percent.round(2)  # округлення до сотих
}).sort_values(by='%', ascending=False)  # сортуємо від найбільш проблемних колонок

print("Пропущені значення (кількість і %):")
print(missing_summary)

Пропущені значення (кількість і %):
                 Пропущені     %
CAPACITY            480386  7.73
FUEL                286351  4.61
N_REG_NEW            67304  1.08
OWN_WEIGHT            4106  0.07
TOTAL_WEIGHT          1953  0.03
VIN                    382  0.01
PERSON                   0  0.00
DEP                      0  0.00
DEP_CODE                 0  0.00
D_REG                    0  0.00
OPER_NAME                0  0.00
OPER_CODE                0  0.00
REG_ADDR_KOATUU         13  0.00
MODEL                    0  0.00
BRAND                    0  0.00
PURPOSE                  0  0.00
BODY                     0  0.00
KIND                     0  0.00
COLOR                    0  0.00
MAKE_YEAR                0  0.00
YEAR                     0  0.00
CAR_AGE                  0  0.00
AGE_CATEGORY            21  0.00


In [47]:
# Перетворюємо КОАТУУ в рядок, щоб уникнути наукової нотації
df_all["REG_ADDR_KOATUU"] = df_all["REG_ADDR_KOATUU"].astype("Int64").astype(str)

# Довідник: перші дві цифри → область
koatuu_to_region = {
    "01": "АР Крим",
    "05": "Вінницька",
    "07": "Волинська",
    "12": "Дніпропетровська",
    "14": "Донецька",
    "18": "Житомирська",
    "21": "Закарпатська",
    "23": "Запорізька",
    "26": "Івано-Франківська",
    "32": "Київська",
    "35": "Кіровоградська",
    "44": "Луганська",
    "46": "Львівська",
    "48": "Миколаївська",
    "51": "Одеська",
    "53": "Полтавська",
    "56": "Рівненська",
    "59": "Сумська",
    "61": "Тернопільська",
    "63": "Харківська",
    "65": "Херсонська",
    "68": "Хмельницька",
    "71": "Черкаська",
    "73": "Чернівецька",
    "74": "Чернігівська",
    "80": "м. Київ",
    "85": "м. Севастополь",
}

# Витягаємо перші дві цифри з коду
df_all["REGION_CODE"] = df_all["REG_ADDR_KOATUU"].str[:2]

# Додаємо назву області
df_all["REGION"] = df_all["REGION_CODE"].map(koatuu_to_region)

In [62]:
# Виводимо унікальні значення REGION
unique_ops = df_all['REGION'].unique()

print("Унікальні операції (REGION):")
for op in unique_ops:
    print(op)

Унікальні операції (REGION):
Дніпропетровська
Київська
Рівненська
м. Київ
Хмельницька
Одеська
Донецька
nan
Івано-Франківська
Харківська
Львівська
Черкаська
Житомирська
Херсонська
Сумська
Полтавська
Закарпатська
Чернівецька
Тернопільська
Запорізька
Чернігівська
Кіровоградська
Миколаївська
Луганська
м. Севастополь


**Висновки із аналізу пропущених значень:**
- CAPACITY (7.73%) – важлива характеристика для аналізу трендів, тому пропуски заповнюємо NaN.
- FUEL (4.61%) – важлива характеристика для аналізу трендів, тому пропуски заповнюємо NaN.
- N_REG_NEW (1.08%) – мало, залишаємо як є, оскілька державний номерний знак ТЗ не є суттєвими для аналізу трендів.
- в інших відсоток пропущених значень малий, можемо ігнорувати.

In [38]:
# Замінюємо порожні рядки або пробіли на NaN у колонці FUEL і CAPACITY
df_all['FUEL'] = df_all['FUEL'].replace(r'^\s*$', np.nan, regex=True)
df_all['CAPACITY'] = df_all['CAPACITY'].replace(r'^\s*$', np.nan, regex=True)

# Перевірка
print(df_all[['FUEL','CAPACITY']].isna().sum())

FUEL        286351
CAPACITY    480386
dtype: int64


In [39]:
# Виводимо унікальні значення OPER_NAME
unique_ops = df_all['OPER_NAME'].unique()

print("Унікальні операції (OPER_NAME):")
for op in unique_ops:
    print(op)

Унікальні операції (OPER_NAME):
ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ЗА ДОГ. КУПIВЛI-ПРОДАЖУ (СГ)
НАЛЕЖНИЙ КОРИСТУВАЧ. РЕЄСТРАЦІЯ
ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВ. ВЛАСН. ПО ДОГОВОРУ УКЛАДЕНОМУ В ТСЦ
ПЕРЕРЕЄСТРАЦІЯ ПРИ ЗАМІНІ НОМЕРНОГО ЗНАКУ
ПЕРЕРЕЄСТРАЦIЯ ПРИ ВТРАТІ СВIДОЦТВА ПРО РЕЄСТРАЦIЮ
ПЕРЕРЕЄСТРАЦІЯ ТЗ ПРИ ВИДАЧІ НОМЕРНОГО ЗНАКУ ВІДПОВІДНОЇ КОМБІНАЦІЇ ЦИФР
ПЕРЕРЕЄСТРАЦIЯ ПРИ ПЕРЕОБЛАДНАННI ДЛЯ РОБОТИ НА ГАЗОВОМУ ПАЛЬНОМУ
ВИДАЧА СВ. ПРО РЕЄСТР ДЛЯ ПОЇЗДКИ ЗА КОРДОН
ЗНЯТТЯ З ОБЛІКУ ПРИ ПОВЕРНЕННІ ІЗ ПОЇЗДКИ ЗА КОРДОН ПО ДОРУЧ.
ПЕРЕРЕЄСТРАЦІЯ ТЗ, В ЗВ’ЯЗКУ ІЗ НАНЕСЕННЯ СПЕЦІАЛЬНИХ ІНДИВІДУАЛЬНИХ НОМЕРІВ АБО ДУБЛЮВАННЯ №КУЗОВА
ЗНЯТТЯ З ОБЛІКУ У ЗВ`ЯЗКУ З ВИБРАКОВКОЮ
ВИДАЧА ТИМЧАСОВОГО РЕЄСТР ТАЛОНУ ПО ДОРУЧЕННЮ ЧИ ІНШОГО ДОКУМ.ПРАВОКОРИСТУВАННЯ 
ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВОГО ВЛАСНИКА ЗА АКТОМ ПРИЙОМУ-ПЕРЕДАЧІ
ПЕРЕРЕЄСТРАЦIЯ У ЗВ`ЯЗКУ ЗI ЗМIНОЮ АНКЕТНИХ ДАНИХ ВЛАСНИКА(БЕЗ ЗМІНИ РЕГІОНУ РЕЄСТРАЦІЇ)
ПЕРЕРЕЄСТРАЦIЯ ТЗ НА НОВ. ВЛАС. ПО УСПАДКУВАННЮ ЗА ЗАКОНОМ ЧИ ЗНЯТТЯМ З ОБЛІКУ В ОРГАНАХ СОЦЗАХИСТУ


**Висновок на основі проведеного вище налізу:**
для аналізу трендів ринку купівлі-продажу авто потрібно залишити тільки ті операції, які відображають факт зміни власника через продаж/купівлю (тобто ринкову угоду).
* Відкидаю: первинну реєстрацію нового авто (це продаж, але від дилера → інший аналіз), операції дарування, спадкування, соціальні передачі, судові рішення, зняття з обліку, технічні заміни, спецумови (гуманітарка, військовий стан).
* Залишаю: операції купівлі-продажу, лізингу, аукціону/торгів, фінансових угод, де є перехід права власності.
* Для аналізу трендів операцій купівлі-продажу залишаються наступні операції:
- ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВ. ВЛАСН. ПО ДОГОВОРУ УКЛАДЕНОМУ В ТСЦ
- ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ЗА ДОГ. КУПІВЛІ-ПРОДАЖУ (СГ)
- ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ЗА ДОГ. КУПІВЛІ-ПРОДАЖУ, УКЛАДЕНОГО У СГ НА ТЗ ЗАРЕЄСТРОВАНИЙ ЗА СГ
- ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВОГО ВЛАСНИКА ЗА ДОГОВОРОМ ФІНАНСОВОГО ЛІЗИНГУ ТА ДКП ПРЕДМЕТА ЛІЗИНГУ
- ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВОГО ВЛАСНИКА ЗА АКТОМ ПРИЙОМУ-ПЕРЕДАЧІ
- ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ЗА ДОГ. КУП.-ПРОД.ТА ІНШ.(ОФОРМ.НОТАРІАЛЬНО)
- ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ТЗ, ПРИДБАНОГО НА АУКЦІОНІ, ЕЛЕКТРОННИХ ТОРГАХ, ТОВАРНІЙ БІРЖІ
- ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ЗА ДОГОВОРОМ МІНИ
- ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВОГО ВЛАСНИКА ПО ДОГОВОРУ КУПІВЛІ-ПРОДАЖУ УКЛАДЕНОМУ В ПРИСУТНОСТІ АДМІНІСТРАТОРА ЦНАП
- ВТОРИННА РЕЄСТРАЦІЯ ТЗ ПО ДОГОВОРУ УКЛАДЕНОМУ В ТСЦ
- РЕЄСТРАЦІЯ ТЗ ПРИДБАНОГО НА АУКЦІОНІ, ЕЛЕКТРОННИХ ТОРГАХ, ТОВАРНІЙ БІРЖІ

In [49]:
# Виводимо перших 5 рядків
df_all.head()

PERSON REG_ADDR_KOATUU  OPER_CODE  \
0      P      1225284101        308   
1      P      3220880905        254   
2      P      5610700000        315   
3      P      8036600000        315   
4      P      5610100000        410   

                                           OPER_NAME       D_REG  DEP_CODE  \
0  ПЕРЕРЕЄСТРАЦІЯ НА НОВОГО ВЛАСНИКА ЗА ДОГ. КУПI...  21.12.2022     12537   
1                    НАЛЕЖНИЙ КОРИСТУВАЧ. РЕЄСТРАЦІЯ  07.09.2022     10000   
2  ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВ. ВЛАСН. ПО ДОГОВОРУ У...  09.12.2022     12286   
3  ПЕРЕРЕЄСТРАЦІЯ ТЗ НА НОВ. ВЛАСН. ПО ДОГОВОРУ У...  02.07.2022     12294   
4          ПЕРЕРЕЄСТРАЦІЯ ПРИ ЗАМІНІ НОМЕРНОГО ЗНАКУ  13.09.2022     12337   

                    DEP       BRAND    MODEL                VIN  ...  \
0              ТСЦ 1248         KIA     SOUL  KNAJT811AC7443261  ...   
1  OLD_ДДАІ МВС УКРАЇНИ      NISSAN  QASHQAI  SJNFBNJ10U2471634  ...   
2              ТСЦ 3247  VOLKSWAGEN     GOLF  WVWZZZ1KZCW268678  ...   
3              ТСЦ 8045     RENAULT  FLUENCE  VF1LZBB0647079727  ...   
4              ТСЦ 5641      NISSAN    TIIDA  3N1FCAC11UK561776  ...   

   OWN_WEIGHT TOTAL_WEIGHT N_REG_NEW  YEAR CAR_AGE AGE_CATEGORY  \
0        1217         1690  АЕ7265ХА  2022      10         6-10   
1        1310         1960  АІ1954ЕМ  2022      10         6-10   
2        1020         1850  АІ2805ОН  2022      10         6-10   
3        1280         1680  КА1560ІЕ  2022      10         6-10   
4        1268         1715  ВК2037ІВ  2022      10         6-10   

   REG_ADDR_KOATUU_str  REGION_CODE  REGION_NAME            REGION  
0           1225284101           12    Львівська  Дніпропетровська  
1           3220880905           32          NaN          Київська  
2           5610700000           56          NaN        Рівненська  
3           8036600000           80          NaN           м. Київ  
4           5610100000           56          NaN        Рівненська  

[5 rows x 27 columns]

In [52]:
# Список технічних стовпців для видалення
cols_to_drop = ['REG_ADDR_KOATUU', 'OPER_CODE', 'OPER_NAME', 'DEP_CODE', 'DEP', 'PURPOSE', 'REG_ADDR_KOATUU_str', 'REGION_CODE', 'REGION_NAME']

# Видаляємо колонки, які не потрібні для аналізу
df_all = df_all.drop(columns=cols_to_drop, errors='ignore')

# Перевірка, які колонки залишились
print("Колонки після видалення технічних стовпців:")
print(df_all.columns)

# Перегляд перших 3 рядків
df_all.head(3)

Колонки після видалення технічних стовпців:
Index(['PERSON', 'D_REG', 'BRAND', 'MODEL', 'VIN', 'MAKE_YEAR', 'COLOR',
       'KIND', 'BODY', 'FUEL', 'CAPACITY', 'OWN_WEIGHT', 'TOTAL_WEIGHT',
       'N_REG_NEW', 'YEAR', 'CAR_AGE', 'AGE_CATEGORY', 'REGION'],
      dtype='object')


PERSON       D_REG       BRAND    MODEL                VIN  MAKE_YEAR  \
0      P  21.12.2022         KIA     SOUL  KNAJT811AC7443261       2012   
1      P  07.09.2022      NISSAN  QASHQAI  SJNFBNJ10U2471634       2012   
2      P  09.12.2022  VOLKSWAGEN     GOLF  WVWZZZ1KZCW268678       2012   

        COLOR      KIND       BODY            FUEL  CAPACITY  OWN_WEIGHT  \
0  КОРИЧНЕВИЙ  ЛЕГКОВИЙ  УНІВЕРСАЛ  БЕНЗИН АБО ГАЗ    1591.0        1217   
1       БІЛИЙ  ЛЕГКОВИЙ    ХЕТЧБЕК  БЕНЗИН АБО ГАЗ    1997.0        1310   
2       БІЛИЙ  ЛЕГКОВИЙ    ХЕТЧБЕК          БЕНЗИН    1390.0        1020   

   TOTAL_WEIGHT N_REG_NEW  YEAR  CAR_AGE AGE_CATEGORY            REGION  
0          1690  АЕ7265ХА  2022       10         6-10  Дніпропетровська  
1          1960  АІ1954ЕМ  2022       10         6-10          Київська  
2          1850  АІ2805ОН  2022       10         6-10        Рівненська

In [54]:
# Показати всі унікальні значення
print(df_all["FUEL"].unique())

# Рахуємо скільки разів зустрічається кожне значення
print(df_all["FUEL"].value_counts())

['БЕНЗИН АБО ГАЗ' 'БЕНЗИН' 'ДИЗЕЛЬНЕ ПАЛИВО' nan 'ЕЛЕКТРО АБО БЕНЗИН'
 'БЕНЗИН, ГАЗ АБО ЕЛЕКТРО' 'ДИЗЕЛЬНЕ ПАЛИВО АБО ГАЗ' '.' 'ЕЛЕКТРО'
 'НЕ ВИЗНАЧЕНО' 'ГАЗ' 'ВІДСУТНЄ' 'ЕЛЕКТРО АБО ДИЗЕЛЬНЕ ПАЛИВО'
 'ГАЗ ТА ЕЛЕКТРО']
FUEL
БЕНЗИН                         2393706
ДИЗЕЛЬНЕ ПАЛИВО                2210472
БЕНЗИН АБО ГАЗ                  982412
ЕЛЕКТРО                         196067
ЕЛЕКТРО АБО БЕНЗИН              102659
ГАЗ                              21554
ЕЛЕКТРО АБО ДИЗЕЛЬНЕ ПАЛИВО      17038
БЕНЗИН, ГАЗ АБО ЕЛЕКТРО           3870
НЕ ВИЗНАЧЕНО                       438
ДИЗЕЛЬНЕ ПАЛИВО АБО ГАЗ            298
ВІДСУТНЄ                           171
.                                  107
ГАЗ ТА ЕЛЕКТРО                      41
Name: count, dtype: int64


In [59]:
# Список значень, які треба вважати відсутніми
invalid_fuel_values = ["НЕ ВИЗНАЧЕНО", "ВІДСУТНЄ", "."]

# Замінюємо їх на NaN
df_all["FUEL"] = df_all["FUEL"].replace(invalid_fuel_values, np.nan)

In [60]:
# Показати всі унікальні значення
print(df_all["FUEL"].unique())

# Рахуємо скільки разів зустрічається кожне значення
print(df_all["FUEL"].value_counts())

['БЕНЗИН АБО ГАЗ' 'БЕНЗИН' 'ДИЗЕЛЬНЕ ПАЛИВО' nan 'ЕЛЕКТРО АБО БЕНЗИН'
 'БЕНЗИН, ГАЗ АБО ЕЛЕКТРО' 'ДИЗЕЛЬНЕ ПАЛИВО АБО ГАЗ' 'ЕЛЕКТРО' 'ГАЗ'
 'ЕЛЕКТРО АБО ДИЗЕЛЬНЕ ПАЛИВО' 'ГАЗ ТА ЕЛЕКТРО']
FUEL
БЕНЗИН                         2393706
ДИЗЕЛЬНЕ ПАЛИВО                2210472
БЕНЗИН АБО ГАЗ                  982412
ЕЛЕКТРО                         196067
ЕЛЕКТРО АБО БЕНЗИН              102659
ГАЗ                              21554
ЕЛЕКТРО АБО ДИЗЕЛЬНЕ ПАЛИВО      17038
БЕНЗИН, ГАЗ АБО ЕЛЕКТРО           3870
ДИЗЕЛЬНЕ ПАЛИВО АБО ГАЗ            298
ГАЗ ТА ЕЛЕКТРО                      41
Name: count, dtype: int64


In [61]:
# Збереження файлу CSV на комп’ютер
output_file = "CARS1_Trends_22_24.csv"
df_all.to_csv(output_file, index=False, encoding='utf-8-sig')

print(f"Файл збережено: {output_file}")

Файл збережено: CARS1_Trends_22_24.csv


In [64]:
# Загальна кількість реєстрацій
total_regs = len(df_all)

# Середня кількість на місяць
regs_per_month = (
    df_all.groupby(df_all['D_REG'].dt.to_period('M'))['D_REG']
    .count()
    .mean()
)

# Найпопулярніший бренд
top_brand = df_all['BRAND'].value_counts().idxmax()

# Найпопулярніший колір
top_color = df_all['COLOR'].value_counts().idxmax()

# Частка електроавто
share_ev = (
    df_all['FUEL'].str.contains('електр', case=False, na=False).sum()
    / len(df_all)
) * 100

# Найпоширеніше пальне
top_fuel = df_all['FUEL'].value_counts().idxmax()

# === Вивід результатів ===
print("Основні KPI")
print(f"Загальна кількість реєстрацій: {total_regs:,}")
print(f"Середня кількість на місяць: {regs_per_month:.0f}")
print(f"Найпопулярніший бренд: {top_brand}")
print(f"Найпопулярніший колір: {top_color}")
print(f"Частка електроавто: {share_ev:.2f}%")
print(f"Найпоширеніше пальне: {top_fuel}")

Основні KPI
Загальна кількість реєстрацій: 6,215,184
Середня кількість на місяць: 145492
Найпопулярніший бренд: VOLKSWAGEN
Найпопулярніший колір: СІРИЙ
Частка електроавто: 5.14%
Найпоширеніше пальне: БЕНЗИН


In [65]:
# Загальна кількість реєстрацій
total_registrations = len(df_all)
print(f"Загальна кількість реєстрацій: {total_registrations}")

# Середня кількість на місяць
# Спершу приводимо колонку D_REG до datetime
df_all['D_REG'] = pd.to_datetime(df_all['D_REG'], errors='coerce')

# Визначаємо період
period_months = ((df_all['D_REG'].max() - df_all['D_REG'].min()).days / 30.44)
avg_per_month = total_registrations / period_months
print(f"Середня кількість реєстрацій на місяць: {avg_per_month:.2f}")

# Найпопулярніший бренд
top_brand = df_all['BRAND'].value_counts().idxmax()
top_brand_count = df_all['BRAND'].value_counts().max()
top_brand_pct = top_brand_count / total_registrations * 100
print(f"Найпопулярніший бренд: {top_brand} ({top_brand_count} реєстрацій, {top_brand_pct:.2f}%)")

# Найпопулярніший колір
top_color = df_all['COLOR'].value_counts().idxmax()
top_color_count = df_all['COLOR'].value_counts().max()
top_color_pct = top_color_count / total_registrations * 100
print(f"Найпопулярніший колір: {top_color} ({top_color_count} реєстрацій, {top_color_pct:.2f}%)")

# Частка електроавто (припускаємо, що KIND або FUEL містить "Electric" чи подібне)
if 'FUEL' in df_all.columns:
    electric_count = df_all[df_all['FUEL'].str.contains('electro|electric', case=False, na=False)].shape[0]
    electric_pct = electric_count / total_registrations * 100
    print(f"Частка електроавто: {electric_count} ({electric_pct:.2f}%)")

# Найпоширеніше пальне
top_fuel = df_all['FUEL'].value_counts().idxmax()
top_fuel_count = df_all['FUEL'].value_counts().max()
top_fuel_pct = top_fuel_count / total_registrations * 100
print(f"Найпоширеніше пальне: {top_fuel} ({top_fuel_count} реєстрацій, {top_fuel_pct:.2f}%)")


Загальна кількість реєстрацій: 6215184
Середня кількість реєстрацій на місяць: 519753.30
Найпопулярніший бренд: VOLKSWAGEN (697880 реєстрацій, 11.23%)
Найпопулярніший колір: СІРИЙ (1870013 реєстрацій, 30.09%)
Частка електроавто: 0 (0.00%)
Найпоширеніше пальне: БЕНЗИН (2393706 реєстрацій, 38.51%)
